python/python_reduction.cc

```diff
@@ -0,0 +1,358 @@
+            opt.m_allow_override, opt.default_value(), false, opt.default_value(), true));
+      else
+        return new py::object(m_py_opt_class(opt.m_name, opt.m_help, opt.m_short_name, opt.m_keep, opt.m_necessary,
+            opt.m_allow_override, py::object(), false, not_supplied, false));
+    }
+  }
+
+  template <typename T>
+  py::object* value_to_pyobject(VW::config::typed_option<std::vector<T>>& opt)
+  {
+    py::list values;
+
+    if (m_opt.was_supplied(opt.m_name))
+    {
+      auto vec = opt.value();
+      if (vec.size() > 0)
+      {
+        for (auto const& opt : vec) { values.append(py::object(opt)); }
+      }
+    }
+
+    return new py::object(m_py_opt_class(opt.m_name, opt.m_help, opt.m_short_name, opt.m_keep, opt.m_necessary,
+        opt.m_allow_override, values, m_opt.was_supplied(opt.m_name), py::list(), opt.default_value_supplied()));
+  }
+
+  template <typename T>
+  py::object* transform_if_t(VW::config::base_option& base_option)
+  {
+    if (base_option.m_type_hash == typeid(T).hash_code())
+    {
+      auto typed = dynamic_cast<VW::config::typed_option<T>&>(base_option);
+      return value_to_pyobject(typed);
+    }
+
+    return nullptr;
+  }
+
+  template <typename TTypes>
+  py::object base_option_to_pyobject(VW::config::base_option& options)
+  {
+    py::object* temp = transform_if_t<typename TTypes::head>(options);
+    if (temp != nullptr)
+    {
+      auto repack = py::object(*temp);
+      delete temp;
+      return repack;
+    }
+
+    return base_option_to_pyobject<typename TTypes::tail>(options);
+  }
+
+  py::object get_vw_option_pyobjects(bool enabled_only)
+  {
+    py::dict dres;
+    auto it = m_option_group_dict.begin();
+
+    while (it != m_option_group_dict.end())
+    {
+      auto reduction_enabled =
+          std::find(m_enabled_reductions.begin(), m_enabled_reductions.end(), it->first) != m_enabled_reductions.end();
+
+      if (((it->first).compare(default_group_name) != 0) && enabled_only && !reduction_enabled)
+      {
+        it++;
+        continue;
+      }
+
+      py::list option_groups;
+
+      for (auto options_group : it->second)
+      {
+        py::list options;
+        for (auto opt : options_group.m_options)
+        {
+          auto temp = base_option_to_pyobject<VW::config::supported_options_types>(*opt.get());
+          options.append(temp);
+        }
+
+        option_groups.append(py::make_tuple(options_group.m_name, options));
+      }
+
+      dres[it->first] = option_groups;
+
+      it++;
+    }
+    return dres;
+  }
+};
+
+class py_cpp_callback
+{
+private:
+  LEARNER::base_learner* base_learner;
+
+  bool m_is_multi = false;
+  multi_ex* examples = nullptr;
+
+  io_buf* model_file = nullptr;
+
+public:
+  py_cpp_callback(LEARNER::base_learner* base_learner) : base_learner(base_learner) {}
+  py_cpp_callback(io_buf* model_file) : model_file(model_file) {}
```

Comment:
either this constructor should initialize base_learner to nullptr or it should be initialized to nullptr along with the rest member vars